src/database/database-sqlite3.cpp

```diff
@@ -606,3 +606,140 @@ void PlayerDatabaseSQLite3::listPlayers(std::vector<std::string> &res)
+{
+}
+
+AuthDatabaseSQLite3::~AuthDatabaseSQLite3()
+{
+	FINALIZE_STATEMENT(m_stmt_read)
+	FINALIZE_STATEMENT(m_stmt_write)
+	FINALIZE_STATEMENT(m_stmt_create)
+	FINALIZE_STATEMENT(m_stmt_delete)
+	FINALIZE_STATEMENT(m_stmt_list_names)
+}
+
+void AuthDatabaseSQLite3::createDatabase()
+{
+	assert(m_database); // Pre-condition
+
+	SQLOK(sqlite3_exec(m_database,
+			      "CREATE TABLE IF NOT EXISTS `auth` ("
+			      "`id` INTEGER PRIMARY KEY AUTOINCREMENT,"
+			      "`name` VARCHAR(32),"
+			      "`password` VARCHAR(512),"
+			      "`privileges` VARCHAR(512),"
+			      "last_login INTEGER);",
+			      NULL, NULL, NULL),
+			"Failed to create auth table");
+}
+
+void AuthDatabaseSQLite3::initStatements()
+{
+	PREPARE_STATEMENT(read, "SELECT id, name, password, privileges, last_login FROM "
+				"auth WHERE name = ?");
+	PREPARE_STATEMENT(write, "UPDATE auth set name = ?, password = ?, privileges = "
+				 "?, last_login = ? WHERE id = ?;");
+	PREPARE_STATEMENT(create, "INSERT INTO auth (name, password, privileges, "
+				  "last_login) VALUES (?, ?, ?, ?);");
+	PREPARE_STATEMENT(delete, "DELETE FROM auth WHERE name = ?");
+	PREPARE_STATEMENT(list_names, "SELECT name FROM auth ORDER BY name DESC");
+}
+
+bool AuthDatabaseSQLite3::getAuth(const std::string &name, AuthEntry &res)
+{
+	verifyDatabase();
+	str_to_sqlite(m_stmt_read, 1, name);
+	if (sqlite3_step(m_stmt_read) != SQLITE_ROW) {
+		sqlite3_reset(m_stmt_read);
+		return false;
+	}
+	res.id = sqlite_to_uint(m_stmt_read, 0);
+	res.name = sqlite_to_string(m_stmt_read, 1);
+	res.password = sqlite_to_string(m_stmt_read, 2);
+	std::string privstring(sqlite_to_string(m_stmt_read, 3));
+	res.privileges = str_split(privstring, ',');
+	res.last_login = (time_t)sqlite_to_uint(m_stmt_read, 4);
+	sqlite3_reset(m_stmt_read);
+	return true;
+}
+
+bool AuthDatabaseSQLite3::saveAuth(const AuthEntry &authEntry)
+{
+	verifyDatabase();
+
+	str_to_sqlite(m_stmt_write, 1, authEntry.name);
+	str_to_sqlite(m_stmt_write, 2, authEntry.password);
+	// TODO: a better 'join', perhaps?
+	std::ostringstream privstream;
+	privstream << *(authEntry.privileges.begin());
+	for (auto p = std::next(authEntry.privileges.begin());
+			p != authEntry.privileges.end(); p++) {
+		privstream << "," << *p;
+	}
+	std::string privstring = privstream.str();
+	str_to_sqlite(m_stmt_write, 3, privstring);
+	int64_to_sqlite(m_stmt_write, 4, authEntry.last_login);
+	int64_to_sqlite(m_stmt_write, 5, authEntry.id);
+	sqlite3_vrfy(sqlite3_step(m_stmt_write), SQLITE_DONE);
+	sqlite3_reset(m_stmt_write);
+
+	return true;
+}
+
+bool AuthDatabaseSQLite3::createAuth(AuthEntry &authEntry)
+{
+	verifyDatabase();
+
+	// id autoincrements
+	str_to_sqlite(m_stmt_create, 1, authEntry.name);
+	str_to_sqlite(m_stmt_create, 2, authEntry.password);
+	// TODO: a better 'join', perhaps?
+	std::ostringstream privstream;
+	privstream << *(authEntry.privileges.begin());
+	for (auto p = std::next(authEntry.privileges.begin());
```

Comment:
with the std::next ? for (auto &p : authEntry.privileges) is sufficient

Reply:
in fact i understand and i know what's the problem. Add a table to store the privileges. This string is a non sense in a SGBD.

Table scheme is:

user_privileges (id, privname, primary key(id, privname))

Reply:
Ok, will take a while though. I'll write some tests first, this is getting bigger than I can check by hand :-P